File: ADMIN/Example-API-Flutter/Example.UserService.WebApp/Services/ProfileService.cs
using Example.Common.Models;
using Example.UserService.WebApp.Models.Profile;
using Example.UserService.WebApp.Repository.IRepository;
using Example.UserService.WebApp.Services.IServices;

namespace Example.UserService.WebApp.Services
{
    /// <summary>
    /// Dịch vụ hồ sơ thí sinh: đọc thông tin user, dựng form sửa, cập nhật profile và đổi mật khẩu.
    /// Phân tách nguồn dữ liệu — đọc qua AppUserRepository, ghi qua AuthenticateRepository (self-service API).
    /// </summary>
    public class ProfileService : IProfileService
    {
        private readonly IAppUserRepository _appUserRepository;
        private readonly IAuthenticateRepository _authenticateRepository;

        // Đọc hồ sơ qua AppUser; cập nhật/đổi mật khẩu qua Authenticate (self-service API).
        public ProfileService(IAppUserRepository appUserRepository, IAuthenticateRepository authenticateRepository)
        {
            _appUserRepository = appUserRepository;
            _authenticateRepository = authenticateRepository;
        }

        /// <summary>Tải hồ sơ thí sinh theo userId từ claims/JWT.</summary>
        public async Task<ProfileViewModel> GetProfileAsync(long userId)
        {
            // Quy tắc nghiệp vụ: userId không hợp lệ thì không gọi API
            if (userId <= 0)
                return new ProfileViewModel { ErrorMessage = "Không xác định được người dùng." };

            var result = await _appUserRepository.GetByIdAsync(userId);
            if (!result.Success || result.Data == null)
            {
                return new ProfileViewModel
                {
                    ErrorMessage = result.Message ?? "Không thể tải thông tin cá nhân."
                };
            }

            return new ProfileViewModel { User = result.Data };
        }

        /// <summary>Dựng form chỉnh sửa từ dữ liệu hồ sơ hiện tại (không cho sửa username).</summary>
        public async Task<EditProfileViewModel> BuildEditFormAsync(long userId)
        {
            var profile = await GetProfileAsync(userId);
            if (profile.User == null)
            {
                return new EditProfileViewModel { ErrorMessage = profile.ErrorMessage };
            }

            var user = profile.User;
            // Chỉ ánh xạ các trường được phép chỉnh sửa lên form
            return new EditProfileViewModel
            {
                Username = user.Username,
                FullName = user.FullName,
                Email = user.Email,
                Phone = user.Phone
            };
        }

        /// <summary>
        /// Cập nhật hồ sơ qua API self-service (JWT xác định user, không cần truyền userId lên API).
        /// Trả về FullName/Email đã trim để Controller cập nhật claims/session.
        /// </summary>
        public async Task<(bool Success, string? ErrorMessage, string? FullName, string? Email)> UpdateProfileAsync(long userId, EditProfileViewModel model)
        {
            if (userId <= 0)
                return (false, "Không xác định được người dùng.", null, null);

            var result = await _authenticateRepository.UpdateProfileAsync(new UpdateProfileModel
            {
                FullName = model.FullName,
                Email = model.Email,
                Phone = model.Phone
            });

            if (!result.Success)
            {
                return (false, result.Message ?? "Cập nhật thông tin thất bại.", null, null);
            }

            return (true, null, model.FullName.Trim(), model.Email.Trim());
        }

        /// <summary>Đổi mật khẩu: API kiểm tra mật khẩu hiện tại và xác nhận mật khẩu mới.</summary>
        public async Task<(bool Success, string? ErrorMessage)> ChangePasswordAsync(long userId, ChangePasswordViewModel model)
        {
            if (userId <= 0)
                return (false, "Không xác định được người dùng.");

            var result = await _authenticateRepository.ChangePasswordAsync(new ChangePasswordModel
            {
                CurrentPassword = model.CurrentPassword,
                NewPassword = model.NewPassword,
                ConfirmPassword = model.ConfirmPassword
            });

            if (!result.Success)
            {
                return (false, result.Message ?? "Đổi mật khẩu thất bại.");
            }

            return (true, null);
        }
    }
}

File: lib/services/profile_service.dart
import '../core/storage/auth_storage.dart';
import '../models/auth/auth_requests.dart';
import '../models/user/app_user.dart';
import '../repositories/app_user_repository.dart';
import '../repositories/auth_repository.dart';
import 'auth_service.dart';

/// Dịch vụ hồ sơ — port từ `ProfileService.cs` (WebApp).
class ProfileService {
  ProfileService(
    this._appUserRepository,
    this._authRepository,
    this._storage,
  );

  final AppUserRepository _appUserRepository;
  final AuthRepository _authRepository;
  final AuthStorage _storage;

  /// Tải hồ sơ theo userId từ session.
  Future<ServiceResult<AppUser>> getProfile(int userId) async {
    if (userId <= 0) {
      return ServiceResult.fail('Không xác định được người dùng.');
    }

    final result = await _appUserRepository.getById(userId);
    if (!result.success || result.data == null) {
      return ServiceResult.fail(
        result.message.isNotEmpty
            ? result.message
            : 'Không thể tải thông tin cá nhân.',
      );
    }

    return ServiceResult.ok(
      result.data!.mergeWithStorage(
        username: _storage.username,
        fullName: _storage.fullName,
        email: _storage.email,
        userType: _storage.userType,
      ),
    );
  }

  /// Cập nhật hồ sơ qua API self-service (JWT xác định user).
  Future<ServiceResult<void>> updateProfile({
    required int userId,
    required String fullName,
    required String email,
    String? phone,
  }) async {
    if (userId <= 0) {
      return ServiceResult.fail('Không xác định được người dùng.');
    }

    final result = await _authRepository.updateProfile(
      UpdateProfileRequest(
        fullName: fullName.trim(),
        email: email.trim(),
        phone: phone?.trim().isEmpty == true ? null : phone?.trim(),
      ),
    );

    if (!result.success) {
      return ServiceResult.fail(
        result.message.isNotEmpty
            ? result.message
            : 'Cập nhật thông tin thất bại.',
      );
    }

    await _storage.updateProfileInfo(
      fullName: fullName.trim(),
      email: email.trim(),
    );

    return ServiceResult.ok();
  }

  /// Đổi mật khẩu.
  Future<ServiceResult<void>> changePassword({
    required int userId,
    required String currentPassword,
    required String newPassword,
    required String confirmPassword,
  }) async {
    if (userId <= 0) {
      return ServiceResult.fail('Không xác định được người dùng.');
    }

    final result = await _authRepository.changePassword(
      ChangePasswordRequest(
        currentPassword: currentPassword,
        newPassword: newPassword,
        confirmPassword: confirmPassword,
      ),
    );

    if (!result.success) {
      return ServiceResult.fail(
        result.message.isNotEmpty
            ? result.message
            : 'Đổi mật khẩu thất bại.',
      );
    }

    return ServiceResult.ok();
  }
}
